### PR TITLE
Initial prototype for creating markdown snippets.

### PR DESCRIPTION
**Pull request by jsuereth**
This is only tests for HTTP semantic conventions.

- Create initial crate that will store the "markdown snippet" feature for semantic conventions
- Outline a public API in the create we can maintain between the CLI and weaver regardless of implementation.
- Write quick & dirty parser in nom for snippet handling
- Add tests for HTTP semconv (metric + span) markdown generation.
- This does NOT wire the code into the CLI yet.

Note:  This is NOT a complete solution, this is just step 1.  There will be following on work to optimise and clean up internal code, as well as fleshing out the full feature set of what SemConv wants to support in markdown snippets.

### CODE REVIEW

File: crates/weaver_semconv_gen/src/gen.rs [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+    fn write_rendered_notes<Out: Write>(&self, out: &mut Out) -> Result<(), Error> {
+        for (counter, note) in self.notes.iter().enumerate() {
+            write!(out, "\n**[{}]:** {}\n", counter + 1, note.trim())?;
+        }
+        Ok(())
+    }
+}
+
+/// Determines an enum's type by the type of its values.
+fn enum_type_string(members: &Vec<EnumEntriesSpec>) -> &'static str {
```
Comment (lquerel) [inline crates/weaver_semconv_gen/src/gen.rs:45]:
```suggestion
fn enum_type_string(members: &[EnumEntriesSpec]) -> &'static str {
```
Clippy should have complain about this one :-)

File: crates/weaver_semconv_gen/src/gen.rs [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+    match members.as_slice() {
+        [first, ..] => match first.value {
```
Comment (lquerel) [inline crates/weaver_semconv_gen/src/gen.rs:47]: I'm a big fan of the Rust pattern matching!

File: crates/weaver_semconv_gen/src/gen.rs [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+            ValueSpec::String(_) => "string",
+        },
+        // TODO - is this a failure scenario?
+        _ => "enum",
+    }
+}
+
+fn write_example_list<Out: Write, Element: std::fmt::Display>(
+    out: &mut Out,
+    list: &Vec<Element>,
```
Comment (lquerel) [inline crates/weaver_semconv_gen/src/gen.rs:59]:
```suggestion
    list: &[Element],
```

File: crates/weaver_semconv_gen/src/gen.rs [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+        ValueSpec::Double(v) => write!(out, "`{v}`")?,
+        ValueSpec::Int(v) => write!(out, "`{v}`")?,
+        ValueSpec::String(v) => write!(out, "`{v}`")?,
+    }
+    Ok(())
+}
+
+fn write_enum_examples_string<Out: Write>(
+    out: &mut Out,
+    members: &Vec<EnumEntriesSpec>,
```
Comment (lquerel) [inline crates/weaver_semconv_gen/src/gen.rs:96]:
```suggestion
    members: &[EnumEntriesSpec],
```

File: crates/weaver_semconv_gen/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,265 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! This crate will generate code for markdown files.
+//! The entire crate is a rush job to catch feature parity w/ existing python tooling by
+//! poorly porting the code into RUST.  We expect to optimise and improve things over time.
+
+#![deny(
```
Comment (lquerel) [inline crates/weaver_semconv_gen/src/lib.rs:7]: These `rustc` and `clippy` directives are now inherited from the main Cargo workspace (thanks to Rust 1.78) so they are redundant I think.

**Review by lquerel — APPROVED** [review]
There is still a lot ToDos, but I understand that this is a first draft. I've made a few minor remarks here and there. Nothing blocking.

File: crates/weaver_semconv_gen/README.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+# Semanticf Convention markdown updater
```
[the remark this answers is not in the document]
Comment (jsuereth) [reply crates/weaver_semconv_gen/README.md]: fixed

File: crates/weaver_semconv_gen/src/gen.rs (hunk 1 above)
Comment (jsuereth) [reply crates/weaver_semconv_gen/src/gen.rs]: Agreed.  Fixed

File: crates/weaver_semconv_gen/src/gen.rs (hunk 4 above)
Comment (jsuereth) [reply crates/weaver_semconv_gen/src/gen.rs:96]: Fixed

File: crates/weaver_semconv_gen/src/lib.rs (hunk 5 above)
Comment (jsuereth) [reply crates/weaver_semconv_gen/src/lib.rs]: Removed

File: crates/weaver_semconv_gen/src/gen.rs (hunk 3 above)
Comment (jsuereth) [reply crates/weaver_semconv_gen/src/gen.rs]: Fixed

File: crates/weaver_semconv_gen/src/gen.rs (hunk 2 above)
Comment (jsuereth) [reply crates/weaver_semconv_gen/src/gen.rs:47]: Agreed! It's the biggest thing I was missing from Scala in other languages, glad to see Rust not only has it, but adopted it in many places.

**Pull request merged**